File: AL/nadongbin/01-greedy/theLawofLargeNumbers/theLawofLargeNumbers.go
package main

import (
	"fmt"
	"sort"
)

func main() {
	n, m, k := 5, 8, 4
	nums := []int{7, 6, 5, 1, 2}
	fmt.Println(solution(n, m, k, nums))
}

func solution(n, m, k int, nums []int) int {
	sort.Slice(nums, func(i, j int) bool { return nums[i] > nums[j] })

	count := (m / (k + 1)) * k
	count += m % (k + 1)

	sum := 0
	sum += count * nums[0]
	sum += (m - count) * nums[1]

	return sum
}

File: AL/nadongbin/01-greedy/theLawofLargeNumbers/theLawofLargeNumbers.ts
function theLawofLargeNumbers(n: number, m: number, k: number, nums: number[]) {
  nums.sort((a, b) => b - a)

  const count = Math.floor(m / (k + 1)) * k + (m % (k + 1))
  const sum = nums[0] * count + nums[1] * (m - count)

  return sum
}

const nums = [7, 6, 5, 1, 2]
const [n, m, k] = [5, 8, 4]

console.log(theLawofLargeNumbers(n, m, k, nums))
